### PR TITLE
Whisper + LLAMA + VITS

### PR DESCRIPTION
**Pull request by SkullMag**
## What's done
I've squeezed whisper, llama and vits code together to have a conversation with tinygrad and also did some refactoring here. It supports all cli parameters, except mmts-tts for now

## TBD
- [x] Refactor vits initialization and add mmts-tts support
- [x] Fix a bug with voice recording for all questions after the first one
- [x] Add configurable llama pre-prompt and examples

### CODE REVIEW

**Comment by SkullMag** [comment]
Now the conversation pipeline is working, so anyone can run this script and have a continuous dialog

**Comment by geohot** [comment]
What's the current latency?

**Comment by SkullMag** [comment]
I'm running on Macbook Air M1 16 GB RAM
It's 15 seconds on average from getting users' input from the mic to playing the whole result. Here is a chat log as an example:
```
User: Hello, how are you?
Stacy: I'm great.
15.339129457999999 seconds
User: They have bipolar disorder.
Stacy: They do, that's right!
16.605268457999998 seconds
User: And do you have a biology soldier?
Stacy: I sure do!
13.694285041999997 seconds
User: What is it?
Stacy: It's a white blood cell that defends you from bacteria and viruses.
15.937712625000003 seconds

**Comment by geohot** [comment]
Running with `PHONEMIZER_ESPEAK_LIBRARY=/opt/homebrew/Cellar/espeak/1.48.04_1/lib/libespeak.dylib PYTHONPATH="." python3 examples/conversation.py --llama_gen 2 --llama_model weights/LLaMA-2/7B`

**Comment by SkullMag** [comment]
Here is a new chat log with a fixed hf weights and quantized tinyllama
```
User: Hello, how are you?
Stacy:  I'm good.
9.274895125

User: What is your next move?
Stacy:  I'm going to a concert.
5.414290707999999

User: Are you a rapper?
Stacy:  No, I like to answer questions.
6.839570375000001

User: But do you have any medical condition?
Stacy:  No, I'm not a doctor.
7.806066375

User: Maybe you are ill?
Stacy:  I think so.
6.116101582999995

User: What soon is do you have?
Stacy:  I'm not sure.
4.206379709000004
```

Latency is much smaller now, but the TTS is still a bottleneck, going to work on optimizing it

**Comment by geohot** [comment]
Getting better!

Running with `PHONEMIZER_ESPEAK_LIBRARY=/opt/homebrew/Cellar/espeak/1.48.04_1/lib/libespeak.dylib PYTHONPATH="." python3 examples/conversation.py --llama_model weights/tinyllama/model.safetensors --llama_tokenizer weights/tinyllama/tokenizer.model`

**Comment by SkullMag** [comment]
Here is an updated with jitted tts:
```
User: What is your name?
Stacy: I'm Stacy
Total: 2.858208000000001

User: Where are you from?
Stacy: I'm from the Bronx, New York
Total: 3.754021625

User: Where do you leave?
Stacy: I leave alone
Total: 2.8882507500000045

User: What is your occupation?
Stacy: I'm an artist
Total: 3.056381000000002

User: Are you a big artist?
Stacy: I'm a rapper with bipolar disorder
Total: 4.710609708999996
```

**Comment by chenyuxyz** [comment]
need to update after recent refactor of vits `ImportError: cannot import name 'download_if_not_present' from 'vits' (/Users/chenyu/code/tinygrad/examples/vits.py)`

File: examples/conversation.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+  return tokenizer_path
+
+def llama_prepare(llama: LLaMa, temperature: float, pre_prompt_path: Path) -> tuple[list[int], str, str, str]:
+  """Prepares a llama model from a specified pre-prompt file"""
+  with open(str(pre_prompt_path)) as f:
+    config = yaml.safe_load(f.read())
+  toks = [llama.tokenizer.bos_id()] + encode_prompt(llama.tokenizer, "system", config["pre_prompt"].replace("\n", " "))
+  for i in config["examples"]:
+    toks += encode_prompt(llama.tokenizer, config["user_delim"], i["user_prompt"])
+    toks += encode_prompt(llama.tokenizer, config["resp_delim"], i["resp_prompt"])
```
Comment (bojanbabic) [inline examples/conversation.py:60]: seems like line duplicate. probably needs removal

**Comment by geohot** [comment]
Congrats! Bounty is yours, and you are a red. Can pay with either PayPal or USDC, reach out to george@tinygrad.org to claim

**Pull request merged**